src/herder/TxSetFrame.cpp

```diff
@@ -331,223 +252,428 @@ TxSetFrame::makeFromTransactions(TxPhases const& txPhases, Application& app,
+        for (int i = 0; i < resolvedTxSet.numPhases(); ++i)
         {
-            valid = valid && txSet->sizeTx(static_cast<Phase>(i)) ==
-                                 outputTxSet->sizeTx(static_cast<Phase>(i));
+            valid =
+                valid && resolvedTxSet.sizeTx(static_cast<Phase>(i)) ==
+                             resolvedOutputTxSet->sizeTx(static_cast<Phase>(i));
         }
     }
-    valid = valid && outputTxSet->checkValid(app, lowerBoundCloseTimeOffset,
-                                             upperBoundCloseTimeOffset);
+
+    valid =
+        valid && resolvedOutputTxSet->checkValid(app, lowerBoundCloseTimeOffset,
+                                                 upperBoundCloseTimeOffset);
     if (!valid)
     {
         throw std::runtime_error("Created invalid tx set frame");
     }
     return outputTxSet;
 }
 
-TxSetFrameConstPtr
-TxSetFrame::makeFromHistoryTransactions(Hash const& previousLedgerHash,
-                                        Transactions const& txs)
-{
-    // We can't use `std::make_shared` here as the constructors are protected.
-    // This may cause leaks in case of exceptions, so keep the constructors
-    // simple and exception-safe.
-    return std::shared_ptr<TxSetFrame>(
-        new TxSetFrame(false, previousLedgerHash, TxSetFrame::TxPhases{txs}));
-}
-
 TxSetFrameConstPtr
 TxSetFrame::makeEmpty(LedgerHeaderHistoryEntry const& lclHeader)
 {
-    // We can't use `std::make_shared` here as the constructors are protected.
-    // This may cause leaks in case of exceptions, so keep the constructors
-    // simple and exception-safe.
     TxPhases phases;
     phases.resize(protocolVersionStartsFrom(lclHeader.header.ledgerVersion,
                                             SOROBAN_PROTOCOL_VERSION)
                       ? static_cast<size_t>(Phase::PHASE_COUNT)
                       : 1);
-    std::shared_ptr<TxSetFrame> txSet(new TxSetFrame(lclHeader, phases));
-    for (int i = 0; i < txSet->mFeesComputed.size(); i++)
+    ResolvedTxSetFrame txSet(lclHeader, phases);
+    for (int i = 0; i < txSet.mFeesComputed.size(); i++)
     {
-        txSet->mFeesComputed[i] = true;
+        txSet.mFeesComputed[i] = true;
     }
-    txSet->computeContentsHash();
-    return txSet;
+    return txSet.toWireTxSetFrame();
 }
 
 TxSetFrameConstPtr
-TxSetFrame::makeFromWire(Application& app, TransactionSet const& xdrTxSet)
+TxSetFrame::makeFromHistoryTransactions(Hash const& previousLedgerHash,
+                                        TxSetFrame::Transactions const& txs)
 {
-    ZoneScoped;
-    std::shared_ptr<TxSetFrame> txSet(new TxSetFrame(
-        false, xdrTxSet.previousLedgerHash, {TxSetFrame::Transactions{}}));
-    size_t encodedSize = xdr::xdr_argpack_size(xdrTxSet);
-    if (!txSet->addTxsFromXdr(app, xdrTxSet.txs, false, std::nullopt,
-                              Phase::CLASSIC))
+    ResolvedTxSetFrame resolvedTxSet(false, previousLedgerHash,
```

Comment:
FWIW 'half-baked' here only means that there is no reference to the parent, and reference to the parent is only needed to avoid code duplication for two getters. So risks are really low here, given that we don't try to call these getters here and frame is then destroyed. Code duplication OTOH might lead to subtle issues in serialization.